File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.java
 class Solution {
    public int maxIceCream(int[] costs, int coins) {

        // Find the maximum cost
        int maxCost = 0;

        for (int cost : costs) {
            maxCost = Math.max(maxCost, cost);
        }

        // count[cost] = number of ice creams having this cost
        int[] count = new int[maxCost + 1];

        for (int cost : costs) {
            count[cost]++;
        }

        int total = 0;

        // Go from cheapest to most expensive
        for (int cost = 1; cost <= maxCost; cost++) {

            if (count[cost] == 0) {
                continue;
            }

            // Cannot afford even one ice cream of this cost
            if (coins < cost) {
                break;
            }

            // Number of ice creams we can buy
            int buyKiya = Math.min(count[cost], coins / cost);

            total += buyKiya;

            coins -= buyKiya * cost;
        }

        return total;
    }
}

File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        sort(begin(costs), end(costs));

        int count = 0;
         
       for(int &costs : costs ){
        if(costs > coins){
            return count;
        }else{
            count++;
            coins -= costs;
        }

       }  
       return count; 
        
    }
};